File: app/src/main/java/com/wafflestudio/snutt_staging/ui/AccountFragment.java
package com.wafflestudio.snutt_staging.ui;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Toast;

import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.login.LoginManager;
import com.facebook.login.LoginResult;
import com.google.common.base.Strings;
import com.wafflestudio.snutt_staging.R;
import com.wafflestudio.snutt_staging.SNUTTBaseActivity;
import com.wafflestudio.snutt_staging.SNUTTBaseFragment;
import com.wafflestudio.snutt_staging.SNUTTUtils;
import com.wafflestudio.snutt_staging.manager.UserManager;
import com.wafflestudio.snutt_staging.model.Facebook;
import com.wafflestudio.snutt_staging.model.SettingsItem;
import com.wafflestudio.snutt_staging.model.User;
import com.wafflestudio.snutt_staging.adapter.SettingsAdapter;
import com.wafflestudio.snutt_staging.view.DividerItemDecoration;

import java.util.ArrayList;
import java.util.List;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

/**
 * Created by makesource on 2017. 1. 24..
 */

public class AccountFragment extends SNUTTBaseFragment {
    private static final String TAG = "ACCOUNT_FRAGMENT";
    private List<SettingsItem> lists;
    private SettingsAdapter adapter;
    private LayoutInflater inflater;
    private CallbackManager callbackManager;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        registerFacebookCallback();

        lists = new ArrayList<>();
        adapter = new SettingsAdapter(lists);
        adapter.setOnItemClickListener(new SettingsAdapter.ClickListener() {
            @Override
            public void onClick(View v, int position) {
                Log.d(TAG, String.valueOf(position) + "-th item clicked!");
                SettingsItem.Type type = lists.get(position).getType();
                switch (type) {
                    case ChangePassword:
                        performChangePassword();
                        break;

                    case ChangeEmail:
                        performChangeEmail();
                        break;

                    case AddIdPassword:
                        performAddIdPassword();
                        break;

                    case LinkFacebook:
                        LoginManager.getInstance().logInWithReadPermissions(AccountFragment.this, null);
                        break;

                    case DeleteFacebook:
                        performDeleteFacebook();
                        break;

                    case Leave:
                        performLeave();
                        break;

                    default: {
                        break;
                    }
                }
            }
        });

        addSettingsList(UserManager.getInstance().getUser());
        adapter.notifyDataSetChanged();
        UserManager.getInstance().getUserInfo(new Callback<User>() {
            @Override
            public void success(final User user, Response response) {
                addSettingsList(user);
                adapter.notifyDataSetChanged();
            }

            @Override
            public void failure(RetrofitError error) {
            }
        });

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_account, container, false);
        RecyclerView recyclerView = (RecyclerView) rootView.findViewById(R.id.account_recyclerView);
        //recyclerView.addItemDecoration(new DividerItemDecoration(getContext(), LinearLayoutManager.VERTICAL));
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(adapter);
        this.inflater = inflater;
        return rootView;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        callbackManager.onActivityResult(requestCode, resultCode, data);
    }

    private void performChangePassword() {
        final View layout = inflater.inflate(R.layout.dialog_change_password, null);

        AlertDialog.Builder alert = new AlertDialog.Builder(getContext());
        alert.setTitle("비밀번호 변경");
        alert.setView(layout);
        alert.setPositiveButton("변경", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                // do nothing in here. because we override this button listener later
            }
        }).setNegativeButton("취소",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int whichButton) {
                        dialog.cancel();
                    }
                });
        final AlertDialog dialog = alert.create();
        dialog.show();
        // change default button handler after dialog show.
        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String oldPassword = ((EditText) layout.findViewById(R.id.now_password)).getText().toString();
                String newPassword = ((EditText) layout.findViewById(R.id.new_password)).getText().toString();
                String newPasswordConfirm = ((EditText) layout.findViewById(R.id.new_password_confirm)).getText().toString();

                if (!newPassword.equals(newPasswordConfirm)) {
                    Toast.makeText(getContext(), "새 비밀번호가 일치하지 않습니다.", Toast.LENGTH_SHORT).show();
                } else {
                    UserManager.getInstance().putUserPassword(oldPassword, newPassword, new Callback() {
                        @Override
                        public void success(Object o, Response response) {
                            Toast.makeText(getContext(), "비밀번호를 변경하였습니다.", Toast.LENGTH_SHORT).show();
                            dialog.dismiss();
                        }

                        @Override
                        public void failure(RetrofitError error) {
                        }
                    });
                }
            }
        });
    }

    private void performChangeEmail() {
        LayoutInflater inflater = (LayoutInflater) getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        final View layout2 = inflater.inflate(R.layout.dialog_change_email, null);
        AlertDialog.Builder alert2 = new AlertDialog.Builder(getContext());
        alert2.setTitle("이메일 변경");
        alert2.setView(layout2);
        alert2.setPositiveButton("변경", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                // do nothing in here. because we override this button listener later
            }
        }).setNegativeButton("취소",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int whichButton) {
                        dialog.cancel();
                    }
                });
        final AlertDialog dialog2 = alert2.create();
        dialog2.show();
        dialog2.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final String email = ((EditText) layout2.findViewById(R.id.email)).getText().toString();
                if (!Strings.isNullOrEmpty(email)) {
                    UserManager.getInstance().putUserInfo(email, new Callback() {
                        @Override
                        public void success(Object o, Response response) {
                            Toast.makeText(getContext(), "이메일 변경에 성공하였습니다.", Toast.LENGTH_SHORT).show();
                            getEmailItem().setDetail(email);
                            adapter.notifyDataSetChanged();
                        }

                        @Override
                        public void failure(RetrofitError error) {
                            Toast.makeText(getContext(), "이메일 변경에 실패하였습니다.", Toast.LENGTH_SHORT).show();
                        }
                    });
                    dialog2.dismiss();
                } else {
                    Toast.makeText(getContext(), "이메일을 입력해주세요.", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private void performAddIdPassword() {
        final View layout = inflater.inflate(R.layout.dialog_add_id, null);

        AlertDialog.Builder alert = new AlertDialog.Builder(getContext());
        alert.setTitle("아이디 비빌번호 추가");
        alert.setView(layout);
        alert.setPositiveButton("추가", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                // do nothing in here. because we override this button listener later
            }
        }).setNegativeButton("취소",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int whichButton) {
                        dialog.cancel();
                    }
                });
        final AlertDialog dialog = alert.create();
        dialog.show();
        // change default button handler after dialog show.
        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final String id = ((EditText) layout.findViewById(R.id.id)).getText().toString();
                final String password = ((EditText) layout.findViewById(R.id.password)).getText().toString();
                final String passwordConfirm = ((EditText) layout.findViewById(R.id.password_confirm)).getText().toString();

                if (!password.equals(passwordConfirm)) {
                    Toast.makeText(getContext(), "비밀번호가 일치하지 않습니다", Toast.LENGTH_SHORT).show();
                } else {
                    UserManager.getInstance().postUserPassword(id, password, new Callback() {
                        @Override
                        public void success(Object o, Response response) {
                            Toast.makeText(getContext(), "아이디를 추가하였습니다", Toast.LENGTH_SHORT).show();
                            updateNewId(id);
                        }

                        @Override
                        public void failure(RetrofitError error) {
                        }
                    });
                    dialog.dismiss();
                }
            }
        });
    }

    private void performDeleteFacebook() {
        AlertDialog.Builder alert = new AlertDialog.Builder(getContext());
        alert.setTitle("페이스북 연동 끊기");
        alert.setMessage("페이스북 연동을 끊겠습니까?");
        alert.setPositiveButton("끊기", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                UserManager.getInstance().deleteUserFacebook(new Callback() {
                    @Override
                    public void success(Object o, Response response) {
                        Toast.makeText(getContext(), "페이스북 연동이 끊어졌습니다", Toast.LENGTH_SHORT).show();
                        LoginManager.getInstance().logOut();
                        updateDeleteFacebook();
                    }

                    @Override
                    public void failure(RetrofitError error) {
                    }
                });
            }
        }).setNegativeButton("취소", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                dialog.cancel();
            }
        });
        AlertDialog dialog = alert.create();
        dialog.show();
    }

    private void performLeave() {
        AlertDialog.Builder alert = new AlertDialog.Builder(getContext());
        alert.setTitle("회원탈퇴");
        alert.setMessage("SNUTT 회원 탈퇴를 하겠습니까?");
        alert.setPositiveButton("회원탈퇴", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                UserManager.getInstance().deleteFirebaseToken(new Callback() {
                    @Override
                    public void success(Object o, Response response) {
                        UserManager.getInstance().deleteUserAccount(new Callback() {
                            @Override
                            public void success(Object o, Response response) {
                                UserManager.getInstance().performLogout();
                                getSNUTTBaseActivity().startIntro();
                                getSNUTTBaseActivity().finishAll();
                            }
                            @Override
                            public void failure(RetrofitError error) {
                                Toast.makeText(getContext(), "회원탈퇴에 실패하였습니다.", Toast.LENGTH_SHORT).show();
                            }
                        });
                    }
                    @Override
                    public void failure(RetrofitError error) {
                        Toast.makeText(getContext(), "회원탈퇴에 실패하였습니다.", Toast.LENGTH_SHORT).show();
                    }
                });
            }
        }).setNegativeButton("취소", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                dialog.cancel();
            }
        });
        AlertDialog dialog = alert.create();
        dialog.show();
    }

    private void updateNewId(String id) {
        int position = -1;
        for (int i = 0;i < lists.size(); i++) {
            if (lists.get(i).getType() == SettingsItem.Type.AddIdPassword) position = i;
        }
        if (position == -1) {
            Log.e(TAG, "Add id & password item not exists!!");
            return;
        }
        lists.remove(position);
        adapter.notifyItemRemoved(position);
        lists.add(position, new SettingsItem("비밀번호 변경", SettingsItem.Type.ChangePassword));
        adapter.notifyItemInserted(position);
        lists.add(position, new SettingsItem("아이디", id, SettingsItem.Type.Id));
        adapter.notifyItemInserted(position);
    }

    private void updateLinkFacebook() {
        int position = -1;
        for (int i = 0;i < lists.size(); i++) {
            if (lists.get(i).getType() == SettingsItem.Type.LinkFacebook) position = i;
        }
        if (position == -1) {
            Log.e(TAG, "link facebook item not exists!!");
            return;
        }
        lists.remove(position);
        adapter.notifyItemRemoved(position);
        lists.add(position, new SettingsItem("페이스북 연동 취소", SettingsItem.Type.DeleteFacebook));
        adapter.notifyItemInserted(position);
        lists.add(position, new SettingsItem("페이스북 이름", "", SettingsItem.Type.FacebookName));
        adapter.notifyItemInserted(position);

        final int pos = position;
        UserManager.getInstance().getUserFacebook(new Callback<Facebook>() {
            @Override
            public void success(Facebook facebook, Response response) {
                getFacebookNameItem().setDetail(facebook.getName());
                adapter.notifyItemChanged(pos);
            }
            @Override
            public void failure(RetrofitError error) {

            }
        });
    }

    private void updateDeleteFacebook() {
        int position = -1;
        for (int i = 0;i < lists.size(); i++) {
            if (lists.get(i).getType() == SettingsItem.Type.FacebookName) position = i;
        }
        if (position == -1) {
            Log.e(TAG, "facebook name item not exists!!");
            return;
        }
        lists.remove(position);
        adapter.notifyItemRemoved(position);
        lists.remove(position);
        adapter.notifyItemRemoved(position);
        lists.add(position, new SettingsItem("페이스북 연동", SettingsItem.Type.LinkFacebook));
        adapter.notifyItemInserted(position);
    }

    private SettingsItem getFacebookNameItem() {
        for (SettingsItem item : lists) {
            if (item.getType() == SettingsItem.Type.FacebookName) return item;
        }
        Log.e(TAG, "facebook name row does not exists!!");
        return null;
    }

    private SettingsItem getEmailItem() {
        for (SettingsItem item : lists) {
            if (item.getType() == SettingsItem.Type.Email) return item;
        }
        Log.e(TAG, "email row does not exists!!");
        return null;
    }

    private void registerFacebookCallback() {
        Log.d(TAG, "register facebook callback called.");
        callbackManager = CallbackManager.Factory.create();
        LoginManager.getInstance().registerCallback(callbackManager, new FacebookCallback<LoginResult>() {
            @Override
            public void onSuccess(LoginResult loginResult) {
                // App code
                String id = loginResult.getAccessToken().getUserId();
                String token = loginResult.getAccessToken().getToken();
                Log.i(TAG, "User ID: " + loginResult.getAccessToken().getUserId());
                Log.i(TAG, "Auth Token: " + loginResult.getAccessToken().getToken());
                UserManager.getInstance().postUserFacebook(id, token, new Callback() {
                    @Override
                    public void success(Object o, Response response) {
                        updateLinkFacebook();
                    }
                    @Override
                    public void failure(RetrofitError error) {

                    }
                });
            }
            @Override
            public void onCancel()
            {
                // App code
                Log.w(TAG, "Cancel");
            }
            @Override
            public void onError(FacebookException exception)
            {
                // App code
                Log.e(TAG, "Error", exception);
            }
        });
    }

    private void addSettingsList(User user) {
        lists.clear();
        if (Strings.isNullOrEmpty(user.getLocal_id())) {
            //lists.add(new SettingsItem(SettingsItem.Type.Header));
            lists.add(new SettingsItem("아이디 비번 추가", user.getLocal_id(), SettingsItem.Type.AddIdPassword));
            lists.add(new SettingsItem(SettingsItem.Type.Header));
            if (Strings.isNullOrEmpty(user.getFb_name())) { // 연동 x
                lists.add(new SettingsItem("페이스북 연동", user.getFb_name(), SettingsItem.Type.LinkFacebook));
            } else {
                lists.add(new SettingsItem("페이스북 이름", user.getFb_name(), SettingsItem.Type.FacebookName));
                lists.add(new SettingsItem("페이스북 연동 취소", SettingsItem.Type.DeleteFacebook));
            }
            lists.add(new SettingsItem(SettingsItem.Type.Header));
            lists.add(new SettingsItem("이메일", user.getEmail(), SettingsItem.Type.Email));
            lists.add(new SettingsItem("이메일 변경", SettingsItem.Type.ChangeEmail));
            lists.add(new SettingsItem(SettingsItem.Type.Header));
            lists.add(new SettingsItem("회원탈퇴", SettingsItem.Type.Leave));
            //lists.add(new SettingsItem(SettingsItem.Type.Header));
        } else {
            //lists.add(new SettingsItem(SettingsItem.Type.Header));
            lists.add(new SettingsItem("아이디", user.getLocal_id(), SettingsItem.Type.Id));
            lists.add(new SettingsItem("비밀번호 변경", SettingsItem.Type.ChangePassword));
            lists.add(new SettingsItem(SettingsItem.Type.Header));
            if (Strings.isNullOrEmpty(user.getFb_name())) { // 연동 x
                lists.add(new SettingsItem("페이스북 연동", SettingsItem.Type.LinkFacebook));
            } else {
                lists.add(new SettingsItem("페이스북 이름", user.getFb_name(), SettingsItem.Type.FacebookName));
                lists.add(new SettingsItem("페이스북 연동 취소", SettingsItem.Type.DeleteFacebook));
            }
            lists.add(new SettingsItem(SettingsItem.Type.Header));
            lists.add(new SettingsItem("이메일", user.getEmail(), SettingsItem.Type.Email));
            lists.add(new SettingsItem("이메일 변경", SettingsItem.Type.ChangeEmail));
            lists.add(new SettingsItem(SettingsItem.Type.Header));
            lists.add(new SettingsItem("회원탈퇴", SettingsItem.Type.Leave));
            //lists.add(new SettingsItem(SettingsItem.Type.Header));
        }
    }

    private SNUTTBaseActivity getSNUTTBaseActivity() {
        return (SNUTTBaseActivity) getActivity();
    }
}

File: app/src/main/java/com/wafflestudio/snutt2/ui/AccountFragment.kt
package com.wafflestudio.snutt2.ui

import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.facebook.CallbackManager
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginManager
import com.facebook.login.LoginResult
import com.google.common.base.Strings
import com.wafflestudio.snutt2.R
import com.wafflestudio.snutt2.SNUTTBaseActivity
import com.wafflestudio.snutt2.SNUTTBaseFragment
import com.wafflestudio.snutt2.adapter.SettingsAdapter
import com.wafflestudio.snutt2.manager.UserManager.Companion.instance
import com.wafflestudio.snutt2.model.SettingsItem
import com.wafflestudio.snutt2.model.User
import java.util.*

/**
 * Created by makesource on 2017. 1. 24..
 */
class AccountFragment : SNUTTBaseFragment() {
    private var lists: MutableList<SettingsItem>? = null
    private var adapter: SettingsAdapter? = null
    private var inflater: LayoutInflater? = null
    private var callbackManager: CallbackManager? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        registerFacebookCallback()
        lists = ArrayList()
        adapter = SettingsAdapter(lists!!)
        adapter!!.setOnItemClickListener(
            object : SettingsAdapter.ClickListener {
                override fun onClick(v: View?, position: Int) {
                    Log.d(TAG, "$position-th item clicked!")
                    val type = lists!!.get(position).type
                    when (type) {
                        SettingsItem.Type.ChangePassword -> performChangePassword()
                        SettingsItem.Type.ChangeEmail -> performChangeEmail()
                        SettingsItem.Type.AddIdPassword -> performAddIdPassword()
                        SettingsItem.Type.LinkFacebook -> LoginManager.getInstance().logInWithReadPermissions(
                            this@AccountFragment,
                            null
                        )
                        SettingsItem.Type.DeleteFacebook -> performDeleteFacebook()
                        SettingsItem.Type.Leave -> performLeave()
                        else -> {
                        }
                    }
                }
            }
        )
        addSettingsList(instance!!.user)
        adapter!!.notifyDataSetChanged()
        instance!!.getUserInfo()
            .bindUi(this,
                onSuccess = {
                    addSettingsList(it)
                    adapter!!.notifyDataSetChanged()
                },
                onError = {
                    // do nothing
                })
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val rootView = inflater.inflate(R.layout.fragment_account, container, false)
        val recyclerView = rootView.findViewById<View>(R.id.account_recyclerView) as RecyclerView
        // recyclerView.addItemDecoration(new DividerItemDecoration(getContext(), LinearLayoutManager.VERTICAL));
        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.adapter = adapter
        this.inflater = inflater
        return rootView
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        callbackManager!!.onActivityResult(requestCode, resultCode, data)
    }

    private fun performChangePassword() {
        val layout = inflater!!.inflate(R.layout.dialog_change_password, null)
        val alert = AlertDialog.Builder(context!!)
        alert.setTitle("비밀번호 변경")
        alert.setView(layout)
        alert.setPositiveButton("변경") { dialog, whichButton ->
            // do nothing in here. because we override this button listener later
        }.setNegativeButton(
            "취소"
        ) { dialog, whichButton -> dialog.cancel() }
        val dialog = alert.create()
        dialog.show()
        // change default button handler after dialog show.
        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
            val oldPassword = (layout.findViewById<View>(R.id.now_password) as EditText).text.toString()
            val newPassword = (layout.findViewById<View>(R.id.new_password) as EditText).text.toString()
            val newPasswordConfirm = (layout.findViewById<View>(R.id.new_password_confirm) as EditText).text.toString()
            if (newPassword != newPasswordConfirm) {
                Toast.makeText(app, "새 비밀번호가 일치하지 않습니다.", Toast.LENGTH_SHORT).show()
            } else {
                instance!!.putUserPassword(oldPassword, newPassword)
                    .bindUi(this,
                        onSuccess = {
                            Toast.makeText(
                                app,
                                "비밀번호를 변경하였습니다.",
                                Toast.LENGTH_SHORT
                            ).show()
                            dialog.dismiss()
                        },
                        onError = {
                            // do nothing
                        })
            }
        }
    }

    private fun performChangeEmail() {
        val inflater = activity!!.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val layout2 = inflater.inflate(R.layout.dialog_change_email, null)
        val alert2 = AlertDialog.Builder(context!!)
        alert2.setTitle("이메일 변경")
        alert2.setView(layout2)
        alert2.setPositiveButton("변경") { dialog, whichButton ->
            // do nothing in here. because we override this button listener later
        }.setNegativeButton(
            "취소"
        ) { dialog, whichButton -> dialog.cancel() }
        val dialog2 = alert2.create()
        dialog2.show()
        dialog2.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
            val email = (layout2.findViewById<View>(R.id.email) as EditText).text.toString()
            if (!Strings.isNullOrEmpty(email)) {
                instance!!.putUserInfo(email)
                    .bindUi(this,
                        onSuccess = {
                            emailItem!!.detail = email
                            adapter!!.notifyDataSetChanged()
                        },
                        onError = {
                            // do nothing
                        }
                    )
                dialog2.dismiss()
            } else {
                Toast.makeText(app, "이메일을 입력해주세요.", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun performAddIdPassword() {
        val layout = inflater!!.inflate(R.layout.dialog_add_id, null)
        val alert = AlertDialog.Builder(context!!)
        alert.setTitle("아이디 비빌번호 추가")
        alert.setView(layout)
        alert.setPositiveButton("추가") { dialog, whichButton ->
            // do nothing in here. because we override this button listener later
        }.setNegativeButton(
            "취소"
        ) { dialog, whichButton -> dialog.cancel() }
        val dialog = alert.create()
        dialog.show()
        // change default button handler after dialog show.
        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
            val id = (layout.findViewById<View>(R.id.id) as EditText).text.toString()
            val password = (layout.findViewById<View>(R.id.password) as EditText).text.toString()
            val passwordConfirm = (layout.findViewById<View>(R.id.password_confirm) as EditText).text.toString()
            if (password != passwordConfirm) {
                Toast.makeText(app, "비밀번호가 일치하지 않습니다", Toast.LENGTH_SHORT).show()
            } else {
                instance!!.postUserPassword(id, password)
                    .bindUi(this,
                        onSuccess = {
                            Toast.makeText(app, "아이디를 추가하였습니다", Toast.LENGTH_SHORT).show()
                            updateNewId(id)
                            dialog.dismiss()
                        },
                        onError = {
                            // do nothing
                        }
                    )
            }
        }
    }

    private fun performDeleteFacebook() {
        val alert = AlertDialog.Builder(context!!)
        alert.setTitle("페이스북 연동 끊기")
        alert.setMessage("페이스북 연동을 끊겠습니까?")
        alert.setPositiveButton("끊기") { dialog, whichButton ->
            instance!!.deleteUserFacebook()
                .bindUi(this,
                    onSuccess = {
                        Toast.makeText(app, "페이스북 연동이 끊어졌습니다", Toast.LENGTH_SHORT).show()
                        LoginManager.getInstance().logOut()
                        updateDeleteFacebook()
                    }, onError = {
                    // do nothing
                })
        }.setNegativeButton("취소") { dialog, whichButton -> dialog.cancel() }
        val dialog = alert.create()
        dialog.show()
    }

    private fun performLeave() {
        val alert = AlertDialog.Builder(context!!)
        alert.setTitle("회원탈퇴")
        alert.setMessage("SNUTT 회원 탈퇴를 하겠습니까?")
        alert.setPositiveButton("회원탈퇴") { dialog, whichButton ->
            val progressDialog = ProgressDialog.show(context, "회원탈퇴", "잠시만 기다려 주세요", true, false)
            instance!!.deleteFirebaseToken()
                .flatMap {
                    instance!!.deleteUserAccount()
                }
                .bindUi(this,
                    onSuccess = {
                        instance!!.performLogout()
                        sNUTTBaseActivity!!.startIntro()
                        sNUTTBaseActivity!!.finishAll()
                        progressDialog.dismiss()
                    },
                    onError = {
                        Toast.makeText(app, "회원탈퇴에 실패하였습니다.", Toast.LENGTH_SHORT).show()
                        progressDialog.dismiss()
                    }
                )
        }.setNegativeButton("취소") { dialog, whichButton -> dialog.cancel() }
        val dialog = alert.create()
        dialog.show()
    }

    private fun updateNewId(id: String) {
        var position = -1
        for (i in lists!!.indices) {
            if (lists!![i].type === SettingsItem.Type.AddIdPassword) position = i
        }
        if (position == -1) {
            Log.e(TAG, "Add id & password item not exists!!")
            return
        }
        lists!!.removeAt(position)
        adapter!!.notifyItemRemoved(position)
        lists!!.add(position, SettingsItem("비밀번호 변경", SettingsItem.Type.ChangePassword))
        adapter!!.notifyItemInserted(position)
        lists!!.add(position, SettingsItem("아이디", id, SettingsItem.Type.Id))
        adapter!!.notifyItemInserted(position)
    }

    private fun updateLinkFacebook() {
        var position = -1
        for (i in lists!!.indices) {
            if (lists!![i].type === SettingsItem.Type.LinkFacebook) position = i
        }
        if (position == -1) {
            Log.e(TAG, "link facebook item not exists!!")
            return
        }
        lists!!.removeAt(position)
        adapter!!.notifyItemRemoved(position)
        lists!!.add(position, SettingsItem("페이스북 연동 취소", SettingsItem.Type.DeleteFacebook))
        adapter!!.notifyItemInserted(position)
        lists!!.add(position, SettingsItem("페이스북 이름", "", SettingsItem.Type.FacebookName))
        adapter!!.notifyItemInserted(position)
        val pos = position
        instance!!.getUserFacebook()
            .bindUi(this,
                onSuccess = {
                    facebookNameItem!!.detail = it.name
                    adapter!!.notifyItemChanged(pos)
                },
                onError = {
                    // do nothing
                })
    }

    private fun updateDeleteFacebook() {
        var position = -1
        for (i in lists!!.indices) {
            if (lists!![i].type === SettingsItem.Type.FacebookName) position = i
        }
        if (position == -1) {
            Log.e(TAG, "facebook name item not exists!!")
            return
        }
        lists!!.removeAt(position)
        adapter!!.notifyItemRemoved(position)
        lists!!.removeAt(position)
        adapter!!.notifyItemRemoved(position)
        lists!!.add(position, SettingsItem("페이스북 연동", SettingsItem.Type.LinkFacebook))
        adapter!!.notifyItemInserted(position)
    }

    private val facebookNameItem: SettingsItem?
        private get() {
            for (item in lists!!) {
                if (item.type === SettingsItem.Type.FacebookName) return item
            }
            Log.e(TAG, "facebook name row does not exists!!")
            return null
        }
    private val emailItem: SettingsItem?
        private get() {
            for (item in lists!!) {
                if (item.type === SettingsItem.Type.Email) return item
            }
            Log.e(TAG, "email row does not exists!!")
            return null
        }

    private fun registerFacebookCallback() {
        Log.d(TAG, "register facebook callback called.")
        callbackManager = CallbackManager.Factory.create()
        LoginManager.getInstance().registerCallback(
            callbackManager,
            object : FacebookCallback<LoginResult> {
                override fun onSuccess(loginResult: LoginResult) {
                    // App code
                    val id = loginResult.accessToken.userId
                    val token = loginResult.accessToken.token
                    Log.i(TAG, "User ID: " + loginResult.accessToken.userId)
                    Log.i(TAG, "Auth Token: " + loginResult.accessToken.token)
                    instance!!.postUserFacebook(id, token)
                        .bindUi(this@AccountFragment,
                            onSuccess = {
                                updateLinkFacebook()
                            },
                            onError = {
                                // do nothing
                            })
                }

                override fun onCancel() {
                    // App code
                    Log.w(TAG, "Cancel")
                    Toast.makeText(app, "페이스북 연동중 오류가 발생하였습니다", Toast.LENGTH_SHORT).show()
                }

                override fun onError(exception: FacebookException) {
                    // App code
                    Log.e(TAG, "Error", exception)
                    Toast.makeText(app, "페이스북 연동중 오류가 발생하였습니다", Toast.LENGTH_SHORT).show()
                }
            }
        )
    }

    private fun addSettingsList(user: User) {
        lists!!.clear()
        if (Strings.isNullOrEmpty(user.local_id)) {
            lists!!.add(SettingsItem(SettingsItem.Type.Header))
            lists!!.add(SettingsItem("아이디 비밀번호 추가", user.local_id, SettingsItem.Type.AddIdPassword))
            lists!!.add(SettingsItem(SettingsItem.Type.Header))
            if (Strings.isNullOrEmpty(user.fb_name)) { // 연동 x
                lists!!.add(SettingsItem("페이스북 연동", user.fb_name, SettingsItem.Type.LinkFacebook))
            } else {
                lists!!.add(SettingsItem("페이스북 이름", user.fb_name, SettingsItem.Type.FacebookName))
                lists!!.add(SettingsItem("페이스북 연동 취소", SettingsItem.Type.DeleteFacebook))
            }
            lists!!.add(SettingsItem(SettingsItem.Type.Header))
            lists!!.add(SettingsItem("이메일", user.email, SettingsItem.Type.Email))
            lists!!.add(SettingsItem("이메일 변경", SettingsItem.Type.ChangeEmail))
            lists!!.add(SettingsItem(SettingsItem.Type.Header))
            lists!!.add(SettingsItem("회원탈퇴", SettingsItem.Type.Leave))
            // lists.add(new SettingsItem(SettingsItem.Type.Header));
        } else {
            lists!!.add(SettingsItem(SettingsItem.Type.Header))
            lists!!.add(SettingsItem("아이디", user.local_id, SettingsItem.Type.Id))
            lists!!.add(SettingsItem("비밀번호 변경", SettingsItem.Type.ChangePassword))
            lists!!.add(SettingsItem(SettingsItem.Type.Header))
            if (Strings.isNullOrEmpty(user.fb_name)) { // 연동 x
                lists!!.add(SettingsItem("페이스북 연동", SettingsItem.Type.LinkFacebook))
            } else {
                lists!!.add(SettingsItem("페이스북 이름", user.fb_name, SettingsItem.Type.FacebookName))
                lists!!.add(SettingsItem("페이스북 연동 취소", SettingsItem.Type.DeleteFacebook))
            }
            lists!!.add(SettingsItem(SettingsItem.Type.Header))
            lists!!.add(SettingsItem("이메일", user.email, SettingsItem.Type.Email))
            lists!!.add(SettingsItem("이메일 변경", SettingsItem.Type.ChangeEmail))
            lists!!.add(SettingsItem(SettingsItem.Type.Header))
            lists!!.add(SettingsItem("회원탈퇴", SettingsItem.Type.Leave))
            // lists.add(new SettingsItem(SettingsItem.Type.Header));
        }
    }

    private val sNUTTBaseActivity: SNUTTBaseActivity?
        private get() = activity as SNUTTBaseActivity?

    companion object {
        private const val TAG = "ACCOUNT_FRAGMENT"
    }
}
